### PR TITLE
[MetaSchedule][M4a] User-API: Tune-TE/TIR/Relay

### PR DESCRIPTION
**Pull request by zxybazh**
This PR is part of the stage M4a of the meta schedule project (#8473).

The architecture is re-designed by Junru and Xiyou. In this PR we introduced the tuning scripts for tir, te & relay workloads with meta schedule. Unittests are added for each type of tuning, and for relay workloads we added tests of ResNet18, MobileNetV2 and BertBase.

Thanks to all co-authors for contributing!

Co-authored-by: Junru Shao <<junrushao1994@gmail.com>>
Co-authored-by: Bohan Hou <<32121147+spectrometerHBH@users.noreply.github.com>>
Co-authored-by: Ruihang Lai <<lairuihangdongdong@qq.com>>
Co-authored-by: Hongyi Jin <<3231950289@qq.com>>
Co-authored-by: Wuwei Lin <<wuwei@apache.org>>
Co-authored-by: Siyuan Feng <<Hzfengsy@sjtu.edu.cn>>

### CODE REVIEW

**Comment by junrushao1994** [comment]
Let's wait for the rest 3 rules to be upstreamed before merging this PR :-)

File: python/tvm/meta_schedule/testing/relay_workload.py [hunk 1]
```diff
@@ -15,13 +15,230 @@
...
+from enum import Enum
 from typing import Dict, Tuple
 
-import tvm.relay.testing  # pylint: disable=unused-import
 from tvm import relay
 from tvm.ir import IRModule
 from tvm.runtime import NDArray
 
+# Model types supported in Torchvision
+class MODEL_TYPE(Enum):  # pylint: disable=invalid-name
```
Comment (junrushao1994) [inline python/tvm/meta_schedule/testing/relay_workload.py:26]: I don't think we should define this huge chunk of models on mainline, at least not for now :-) Let's find alternatives

Comment (zxybazh) [reply python/tvm/meta_schedule/testing/relay_workload.py:27]: Hey, I simplied most of the script, does the script look good to you?

Comment (junrushao1994) [reply python/tvm/meta_schedule/testing/relay_workload.py:27]: Looks great!

File: tests/python/unittest/test_meta_schedule_tune_relay.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        assert np.allclose(actual_output, expected_output, rtol=1e-4, atol=2e-4)
+
+
+if __name__ == """__main__""":
+    test_meta_schedule_tune_relay("resnet18", 1, "llvm --num-cores=16")
+    test_meta_schedule_tune_relay("resnet18", 1, "nvidia/geforce-rtx-3070")
+    test_meta_schedule_tune_relay("mobilenet_v2", 1, "llvm --num-cores=16")
+    test_meta_schedule_tune_relay("mobilenet_v2", 1, "nvidia/geforce-rtx-3070")
+    test_meta_schedule_tune_relay("bert_base", 1, "llvm --num-cores=16")
+    test_meta_schedule_tune_relay("bert_base", 1, "nvidia/geforce-rtx-3070")
```
Comment (junrushao1994) [inline tests/python/unittest/test_meta_schedule_tune_relay.py:151]: Wonder how long would these 6 unittests take in total?

**Review by Hzfengsy — APPROVED** [review]

File: tests/python/unittest/test_meta_schedule_tune_relay.py (hunk 2 above)
Comment (zxybazh) [reply tests/python/unittest/test_meta_schedule_tune_relay.py:151]: Around 30 minutes.

File: src/meta_schedule/task_scheduler/task_scheduler.cc [hunk 3]
```diff
@@ -124,7 +124,7 @@ void TaskSchedulerNode::Tune() {
 
   int running_tasks = tasks.size();
   for (int task_id; (task_id = NextTaskId()) != -1;) {
-    LOG(INFO) << "Scheduler picks Task #" << task_id << ": " << tasks[task_id]->task_name;
+    LOG(INFO) << "Scheduler picks Task #" << task_id + 1 << ": " << tasks[task_id]->task_name;
```
Comment (MasterJH5574) [inline src/meta_schedule/task_scheduler/task_scheduler.cc:127]: Hi Xiyou! I'm curious about this change. After this change the printed ids at several places differ (here with `+ 1` while others without.)

https://github.com/apache/tvm/blob/0c836b73ffd9669bcc416515dce6436cbd7d7ebe/src/meta_schedule/task_scheduler/task_scheduler.cc#L96

https://github.com/apache/tvm/blob/0c836b73ffd9669bcc416515dce6436cbd7d7ebe/src/meta_schedule/measure_callback/echo_statistics.cc#L228

Perhaps it's better to keep them consistent :eyes:?

Comment (zxybazh) [reply src/meta_schedule/task_scheduler/task_scheduler.cc:127]: Hi, thanks for pointing that out. The reason for this change is previously we have `task 0/18` to `task 17/18` which seems confusing to me. I intend to make sure the log is consistent but I might missed a couple places. I'll make a thorough check to update that.

**Pull request merged**